KAM support initial code

susanodom opened the pull request:
Fixes https://github.com/kappnav/issues/issues/100

susanodom committed:
KAM support initial code

susanodom committed:
more clean up with KAM impl

mrglavas commented on src/main/java/application/rest/v1/configmaps/ConfigMapProcessor.java:
```diff
@@ -59,113 +60,141 @@ public ConfigMapProcessor(String kind) {
+                String configMapName = getConfigMapName(type, name, subkind, kind);
+                map = getConfigMap(client, kam, namespace, configMapName, builder);
+            } else { // status mapping configmap       
+                if (type == ConfigMapType.STATUS_MAPPING && builder.getConfigMap().entrySet().size() == 0) {
```
How is the status config map computed after the introduction of KAM? I may have missed something, but it looks like the ConfigMapProcessor is mapping all status to "unregistered" and not the specific status config maps for the component kind.

mrglavas commented on src/main/java/application/rest/v1/configmaps/ConfigMapProcessor.java:
```diff
@@ -174,20 +203,52 @@ private ConflictAction getConflictAction(JsonObject map) {
                     String s = e.getAsString();
                     if ("merge".equals(s)) {
+                        if (Logger.isDebugEnabled()) 
+                            Logger.log(className, "getConflictAction", Logger.LogType.DEBUG, "MERGE");
```
Wonder if this logging message needs more text to be useful, like the name/namespace of the map that the value was read from.

mrglavas commented on src/main/java/application/rest/v1/configmaps/ConfigMapProcessor.java:
```diff
@@ -174,20 +203,52 @@ private ConflictAction getConflictAction(JsonObject map) {
                     }
                     else if ("replace".equals(s)) {
+                        if (Logger.isDebugEnabled()) 
+                            Logger.log(className, "getConflictAction", Logger.LogType.DEBUG, "REPLACE");
```
Same comment as above.

mrglavas commented on src/main/java/application/rest/v1/configmaps/ConfigMapProcessor.java:
```diff
@@ -174,20 +203,52 @@ private ConflictAction getConflictAction(JsonObject map) {
+
         // Default value
+        if (Logger.isDebugEnabled()) 
+            Logger.log(className, "getConflictAction", Logger.LogType.DEBUG, "MERGE");
```
Same comment as above.

mrglavas commented on src/main/java/application/rest/v1/configmaps/ConfigMapProcessor.java:
```diff
@@ -174,20 +203,52 @@ private ConflictAction getConflictAction(JsonObject map) {
+                if (getConflictAction(cMap) == ConflictAction.REPLACE) {
+                    if (Logger.isDebugEnabled()) 
+                        Logger.log(className, "getConflictAction", Logger.LogType.DEBUG, 
+                                   "Stop merging with a REPLACE conflicit action");
```
correction: conflicit -> conflict.

mrglavas commented on src/main/java/application/rest/v1/configmaps/ConfigMapCache.java:
```diff
@@ -122,6 +124,29 @@ private Selector getSelector() {
+                Logger.log(CLASS_NAME, "getConfigMapsAsJSON", Logger.LogType.DEBUG, 
+                    "configMapsList["+i+"]="+configMapsList.get(i));
+            }                                                   
+            String[] mapName = configMapsList.get(i).split("@");  // list entry format: mapname@namespace
```
The ConfigMapCache uses QName (created for the XML APIs, but really just a pair of strings to represent name/namespace) to avoid having to do this kind of text manipulation.

susanodom replied:
OK - Using QName.

mrglavas commented on src/main/java/application/rest/v1/configmaps/KindActionMappingProcessor.java:
```diff
@@ -0,0 +1,462 @@
+            } else if ( parts[i].equals("${name}") ) {
+                actualMapName = actualMapName + name;
+            } else if ( parts[i].equals("${kind}-${subkind}") ) {
+                actualMapName = actualMapName + kind.toLowerCase() + "-" + subkind.toLowerCase();
```
Should pass Locale.ENGLISH to every call to String.toLowerCase(). You can get some surprising results if you don't.

mrglavas commented on src/main/java/application/rest/v1/configmaps/KindActionMappingProcessor.java:
```diff
@@ -0,0 +1,462 @@
+            } else if ( parts[i].equals("${kind}-${subkind}") ) {
+                actualMapName = actualMapName + kind.toLowerCase() + "-" + subkind.toLowerCase();
+            } else if ( parts[i].equals("${kind}") ) {
+                actualMapName = actualMapName + kind.toLowerCase();
```
Should pass Locale.ENGLISH to every call to String.toLowerCase(). You can get some surprising results if you don't.

mrglavas commented on src/main/java/application/rest/v1/configmaps/KindActionMappingProcessor.java:
```diff
@@ -0,0 +1,462 @@
+                                Logger.log(className, "getConfigMapsFromKAMs", Logger.LogType.DEBUG,
+                                           "\nKindActionMapping found: " + item);
+
+                            if (item != null) {
```
You already checked for null above.

mrglavas commented on src/main/java/application/rest/v1/configmaps/KindActionMappingProcessor.java:
```diff
@@ -0,0 +1,462 @@
+     * @param kind
+     * @return KSN, KS, KN, or K 
+     */
+    private int exemineMappingProperties(String name, String subkind, String kind) {
```
Pretty sure you meant to call this method examineMappingProperties().

mrglavas commented on src/main/java/application/rest/v1/configmaps/KindActionMappingProcessor.java:
```diff
@@ -0,0 +1,462 @@
+     * @param gKind
+     * @return true if found; otherwise false
+     */
+    private boolean foundMatchedConfigMap(int cNumFields, int kNumFields,
```
foundMatchedConfigMap() -> findMatchedConfigMap()?

mrglavas reviewed (changes requested):
See review comments. I didn't review KindActionMappingProcessor too deeply because I'm not familiar enough with the design. Chris will be much better for reviewing its correctness.

cvignola commented on src/main/java/application/rest/v1/configmaps/KindActionMappingProcessor.java:
```diff
@@ -0,0 +1,462 @@
+                                                           "\nmapname = " + mapname);
+
+                                                if (isApiVersionMatch(apiVersion, compApiVersion)) {
+                                                    int compPropsIdx = exemineMappingProperties(compName, compSubkind, compKind);
```
Nit:  function name spelling "exemineMapping..." should be "examineMapping..."

cvignola commented on src/main/java/application/rest/v1/configmaps/KindActionMappingProcessor.java:
```diff
@@ -0,0 +1,462 @@
+                                String kamNamespace = getKAMNamespace(item);
+                                if (element != null) {
+                                    JsonObject spec = element.getAsJsonObject();
+                                    int precedenceIndex = spec.get(PRECEDENCE_PROPERTY_NAME).getAsInt()-1;
```
Spec.precedence is optional.  Must handle default value of 1.  See https://github.com/kappnav/design/blob/master/kind-action-mapping.md#kindactionmapping-custom-resource-definition

cvignola commented on src/main/java/application/rest/v1/configmaps/KindActionMappingProcessor.java:
```diff
@@ -0,0 +1,462 @@
+                String comp_grp_version[] = compApiVersion.split("/");
+
+                if ( (grp_version.length == 2) && (comp_grp_version.length == 2) ){ // group/version - wildcard match
+                    if (grp_version[0].equals(comp_grp_version[0]) || (grp_version[0].equals(WILDCARD)) )
```
Suggestion: index constants for 0 and 1 would make this more readable - e.g.  GROUP=0, VERSION=1

cvignola commented on src/main/java/application/rest/v1/configmaps/KindActionMappingProcessor.java:
```diff
@@ -0,0 +1,462 @@
+                match = true;
+            } else {
+                String grp_version[] = apiVersion.split("/");
+                String comp_grp_version[] = compApiVersion.split("/");
```
Might be good to add a DEBUG log entry to display the split results.  If grp_version and comp_grp_version array lengths differ, this function ends silently with match=false

cvignola commented on src/main/java/application/rest/v1/configmaps/KindActionMappingProcessor.java:
```diff
@@ -0,0 +1,462 @@
+                            match = true;
+                        }
+                } else if ( (grp_version.length == 1) && (comp_grp_version.length == 1) ) { // verion only
+                    if (grp_version[0].equals(comp_grp_version[0]) || (grp_version[0].equals(WILDCARD)) ) {
```
If you use index constants,  0 is not GROUP in this case, it's actually VERSION.  So maybe use a different constant here, such as NO_GROUP_VERSION or GROUPLESS_VERSION = 0, since this is the case for those older, built-in kinds that have version, but no group.

cvignola commented on src/main/java/application/rest/v1/configmaps/KindActionMappingProcessor.java:
```diff
@@ -0,0 +1,462 @@
+    /**
+     * Find a matched configmap defined in a given KAM 
+     * 
+     * @param cNumFields
```
Suggestion:  add a description for each @param - reader has to keep jumping back and forth to where this method is called to figure out what each parameter is.

cvignola commented on src/main/java/application/rest/v1/configmaps/KindActionMappingProcessor.java:
```diff
@@ -0,0 +1,462 @@
+                Logger.log(className, "foundMatchedConfigMap", Logger.LogType.ENTRY, 
+                           "cNumField = " + cNumFields + " kNumField = " + kNumFields);
+
+        if ( (cNumFields == -1) || (kNumFields == -1)) {
```
This is kind of an error case, isn't it?  Do you maybe want to log also an error to make it more prominent?

cvignola commented on src/main/java/application/rest/v1/configmaps/KindActionMappingProcessor.java:
```diff
@@ -0,0 +1,462 @@
+        boolean match = false;
+        if (cNumFields == kNumFields) {
+            switch (kNumFields) {
+            case KSN:
```
A bit of commentary in here would go a long way in helping the reader.  E.g. this case is KAM specifies KSN and tests if the resource's KSN value matches (exactly or via wildcard), the KAM mapping's KSN value.

cvignola commented on src/main/java/application/rest/v1/configmaps/KindActionMappingProcessor.java:
```diff
@@ -0,0 +1,462 @@
+                                                if (isApiVersionMatch(apiVersion, compApiVersion)) {
+                                                    int compPropsIdx = exemineMappingProperties(compName, compSubkind, compKind);
+                                                    int kamMappingPropIdx = exemineMappingProperties(name, subkind, kind);
+                                                    boolean found = foundMatchedConfigMap(compPropsIdx, kamMappingPropIdx, 
```
Maybe use a clearer method name. "Found" is past tense:  the configmap is not found yet.  In fact, the method does not examine configmaps at all.  What the method is trying to do is determine if the subject resource maps to the current KAM mapping rule.   A method name that more specifically identifies the task might be:   

```
boolean matches= resourceMatchesRule(....)
```

cvignola commented on src/main/java/application/rest/v1/configmaps/KindActionMappingProcessor.java:
```diff
@@ -0,0 +1,462 @@
+                    if ((gKind.equals(cKind)) || (gKind.equals(WILDCARD)))
+                        match = true;
+            }
+        } else if (kNumFields == K) {
```
Suggestion: might want to switch order of test done on 381 with the test done on 384.  I know you get the same outcome either way.  It's just that if you change the order,  the order of condition in this method would match the hierarchy order - i.e. KSN, KS, K - and therefore be more intuitive to the reader and raise less questions.  After I saw the order of the condition tests in this method of 'KSN, K, KS' it made me concerned the order was wrong and I had to study the caller of this method to determine the order of testing the conditions doesn't actually matter.

cvignola commented on src/main/java/application/rest/v1/configmaps/KindActionMappingProcessor.java:
```diff
@@ -0,0 +1,462 @@
+        this.compKind = kind;
+    }
+
+    /**
```
Would probably be a good idea to document the hierarchy details in a block comment somewhere in here.

cvignola commented on src/main/java/application/rest/v1/configmaps/KindActionMappingProcessor.java:
```diff
@@ -0,0 +1,462 @@
+        if (Logger.isEntryEnabled()) 
+                Logger.log(className, "getConfigMapsFromKAMs", Logger.LogType.ENTRY,"");
+
+        String[][] mapNamesFound = new String[MAX_PRECEDENCE][TOTAL_INDIVIDUAL_MAPPINGS];
```
This implementation does not handle multiple KAMs that have the same precedence level.   The spec says: 


_"KindActionMappings resources with the same precedence value are processed together in arbitrary order."_

E.g. need a structure something like this: 

```
String[MAX_PRECEDENCE][KAM-N][TOTAL_INDIVIDUAL_MAPPINGS] 
```

cvignola commented on src/main/java/application/rest/v1/configmaps/KindActionMappingProcessor.java:
```diff
@@ -0,0 +1,462 @@
+/*
```
I request you document test cases and provide sample KAMs for those test cases.  There needs to be someway to execute those tests to confirm correct behavior.  Store this all somewhere in Git.

kinueng committed:
apiVersion: Add suppose for the percentage symbol

- The apiVersion value will be url encoded, which means the `%` can
exist.  Add the `%` to the regular expression for the apiVersion
query parameter.

kinueng committed:
Merge pull request #71 from kappnav/apiVersion-supportPercentageSymbol

apiVersion: Add suppose for the percentage symbol

susanodom committed:
Code review updates

susanodom committed:
Merge branch 'Issue#100-KindActionMappingSupport' of github.com:kappnav/apis into Issue#100-KindActionMappingSupport
code review updates

mrglavas reviewed (approved):
Looks good now. Thanks.

cvignola commented on src/main/java/application/rest/v1/KAppNavEndpoint.java:
```diff
@@ -163,8 +167,45 @@ public static JsonObject getItemAsObject(ApiClient client, Object resource) {
+        if (Logger.isEntryEnabled())
+            Logger.log(className, "getComponentKind", Logger.LogType.ENTRY,"");
+
+        final JsonElement kind_e = component.get(KIND_PROPRETY_NAME);
```
nit:  spelling should be PROPERTY not PROPRETY

susanodom committed:
fix bad spelling
